--- a/src/SDL_sound_wav.c
+++ b/src/SDL_sound_wav.c
@@ -211,8 +211,20 @@
     Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
     wav_t *w = (wav_t *) internal->decoder_private;
     Uint32 max = (internal->buffer_size < (Uint32) w->bytesLeft) ?
-                    internal->buffer_size : (Uint32) w->bytesLeft;
+                  internal->buffer_size : (Uint32) w->bytesLeft;
 
+    /* We need to convert 24-bit PCM to an SDL-friendly AUDIO_S32SYS ... */
+    if (w->fmt->wBitsPerSample == 24) {
+        const Uint32 num_samples = max / 3;
+
+        /* we're going to expand by 25%...3 bytes to 4. Make sure the buffer has room to expand. */
+        max = (num_samples - (num_samples / 4)) * 3;
+        if (max == 0) {
+            sample->flags |= SOUND_SAMPLEFLAG_EOF;
+            return 0;
+        }
+    }
+
     SDL_assert(max > 0);
 
         /*
@@ -234,6 +246,19 @@
     else if (retval < internal->buffer_size)
         sample->flags |= SOUND_SAMPLEFLAG_EAGAIN;
 
+    /* deal with 24-bit PCM. */
+    if ((retval > 0) && (w->fmt->wBitsPerSample == 24)) {
+        const Uint32 total = retval / 3;
+        const Uint8 *src = ((Uint8 *)internal->buffer + retval) - 3;
+        Uint32 *dst = (Uint32 *) (((Uint8 *)internal->buffer + (total * 4)) - 4);
+        Uint32 i;
+        for (i = 0; i < total; i++, dst--, src -= 3) {
+            const Uint32 sample = ((Uint32) src[0]) | (((Uint32) src[1]) << 8) | (((Uint32) src[2]) << 16);
+            *dst = sample << 8;  /* shift it up so the most significant bits cover the 32-bit space. */
+        }
+        retval = total * 4;
+    }
+
     return retval;
 } /* read_sample_fmt_normal */
 
@@ -662,19 +687,17 @@
     } /* if */
     else
     {
-        if (fmt->wBitsPerSample == 4)
-            sample->actual.format = AUDIO_S16SYS;
-        else if (fmt->wBitsPerSample == 8)
-            sample->actual.format = AUDIO_U8;
-        else if (fmt->wBitsPerSample == 16)
-            sample->actual.format = AUDIO_S16LSB;
-        else if (fmt->wBitsPerSample == 32)
-            sample->actual.format = AUDIO_S32LSB;
-        else
+        switch (fmt->wBitsPerSample)
         {
-            SNDDBG(("WAV: %d bits per sample!?\n", (int) fmt->wBitsPerSample));
-            BAIL_MACRO("WAV: Unsupported sample size.", 0);
-        } /* else */
+            case 4: sample->actual.format = AUDIO_S16SYS; break;
+            case 8: sample->actual.format = AUDIO_U8; break;
+            case 16: sample->actual.format = AUDIO_S16LSB; break;
+            case 24: sample->actual.format = AUDIO_S32SYS; break;
+            case 32: sample->actual.format = AUDIO_S32LSB; break;
+            default:
+                SNDDBG(("WAV: %d bits per sample!?\n", (int) fmt->wBitsPerSample));
+                BAIL_MACRO("WAV: Unsupported sample size.", 0);
+        } /* switch */
     } /* else */
 
     BAIL_IF_MACRO(!read_fmt(rw, fmt), NULL, 0);
